File: api/routes/accounts.js
var Accounts = require('payparrot_models/objects/accounts.js');
var Suscriptions = require('payparrot_models/objects/suscriptions.js');
var Payments = require('payparrot_models/objects/payments.js');
var Plans = require('payparrot_models/objects/plans.js');
var AccountsPlans = require('payparrot_models/objects/accounts_plans.js');
var pp_stats = require('payparrot_models/libs/stats.js');
var _ = require('underscore');
var crypto = require('crypto');
var async = require('async');

var ObjectId = require('mongoose').Types.ObjectId;

exports.create = function(req, res){;
	var account = new Accounts(req.body);
	var current_date = (new Date()).valueOf().toString();
	account.credentials = {
		'public_token': crypto.createHash('sha1').update(current_date + Math.random().toString()).digest('hex'),
		'private_token': crypto.createHash('sha1').update(current_date + Math.random().toString()).digest('hex')
	};
	account.create_password(req.body.password);
	account.save(function(err){
		if(err) res.throw_error(err, 503);
		else{
			res.statusCode = 201;
			res.send({id: account._id});
		}
	}); 
};

function set_stats(account, callback) {
	account.stats = {
		parrots_total: 0,
		parrots_today: 0,
		payments_total: 0,
		payments_today: 0
	}
	async.parallel([
	    async.apply(pp_stats.parrots_total, account),
	    async.apply(pp_stats.parrots_today, account),
		async.apply(pp_stats.payments_total, account),
		async.apply(pp_stats.payments_today, account)
	], function(err, results){
		callback(err, results);
	});
}

exports.get = function(req, res){
	var account_id = req.params.account_id;
	if(account_id == 'me'){
		var user = req.user.returnJSON();
		var account_id = user.id;
	}
	Accounts.findOne({'_id': account_id}, {}, function (err, account){
		if(err) res.throw_error(err, 503);
		else{
			if (account) {
				set_stats(account, function(err_stats, results){
					if(err_stats) res.throw_error(err_stats, 503);
					else{
						res.send(account.returnJSON());
					}
				});
			} else {
				res.throw_error(null, 404);
			}
		}
	});
};

exports.update = function(req, res){
	Accounts.findOne({_id: req.params.account_id}, {}, function (err, account){
		if(err) res.throw_error(err, 503);
		else{
			if (account) {
				_.extend(account,req.body);
				if(req.body.password){
					account.create_password(req.body.password);
				}
				account.save(function(err_save){
					if(err_save) res.throw_error(err_save, 503);
					else{
						res.statusCode = 204;
						res.send();
					}
				});
			} else {
				res.throw_error(null, 404);
			}
		}
	});
}

exports.get_credentials = function(req, res){
	Accounts.findOne({_id: req.params.account_id}, {credentials: 1}, function (err, account){								
		if(err) res.throw_error(err, 503);
		else{
			if (account) {
				res.statusCode = 200;
				res.send(account.credentials);
			} else {
				res.throw_error(null, 404);
			}
		}
	});
};

exports.logged = function(req, res){
	if(req.user.setup){
		if (req.user.subscriptions) {
			res.redirect('/app.html');
		} else {
			res.redirect('/accounts/subscriptions');	
		}		
	} else {
		res.redirect('/accounts/setup');	
	}
	
};

exports.login = function(req, res){
	res.render('login.ejs');
};

exports.signup = function(req, res){
	Accounts.findOne({email: req.body.email},function(err,account){
		if (account) {
			res.render('signedup.ejs',{layout: true, status: false, params: req.body});
		} else {
			var account = new Accounts(req.body);
			var current_date = (new Date()).valueOf().toString();
			account.credentials = {
				'public_token': crypto.createHash('sha1').update(current_date + Math.random().toString()).digest('hex'),
				'private_token': crypto.createHash('sha1').update(current_date + Math.random().toString()).digest('hex')
			};
			account.create_password(req.body.password);
			account.save(function(err){
				if(err) res.throw_error(err, 503);
				else{
					res.render('signedup.ejs',{layout: true, status: true});
				}
			}); 	
		}
	});
	//res.render('signup.ejs');
};

exports.setup = function(req,res) {
	res.render('steps.ejs', {account_id: req.user._id});
}

exports.subscriptions = function(req,res) {
	res.render('subscriptions.ejs', {account_id: req.user._id});
}

exports.token_auth = function(req, res, next){
	var token = req.query.token;
	var account_id = req.query.account_id;
	if(!account_id){
		account_id = req.params.account_id;
	}
	Accounts.findOne({'_id':account_id,'credentials.private_token': req.query.token}, {}, function (err, account){
		if(err) res.throw_error(err, 503);
		else{
			if (account) {
				return next();
			} else {
				// TODO: CHANGE STATUS
				res.throw_error(null, 404);
			}
		}
	});
}

exports.update_plan = function(req,res) {

	var get_plan = function(account, callback){
		Plans.findOne({name: req.body.name}, function(err,plan){
			if(err) callback(err);
			else {
				if (!plan) {
					callback(err, null);
				} else {
					callback(null, account, plan);
				}
			}
		});	
	};
	var create_account_plan = function(account, plan, callback){
		var plan_data = plan.toJSON();
		delete plan_data._id;
		plan_data['active'] = true;
		plan_data['account_id'] = account._id;
		AccountsPlans.findOne({account_id: req.params.account_id, active:1}, function (err, account_plan){
			if(err) callback(err);
			else{
				var new_account_plan = new AccountsPlans(plan_data);
				new_account_plan.save(function(err){
					if(err) callback(err);
					else {
						callback(null, account_plan, new_account_plan);
					}
				});
			}
		});			
	};
	var update_account_plan = function(account_plan, new_account_plan, callback){
		if (account_plan) {
			account_plan.active = false;
			account_plan.save(function(err){
				if (!err) callback(null, new_account_plan);
				else calback(err);
			});
		} else {
			callback(null, new_account_plan);
		}						
	};
	async.waterfall(
		[
			async.apply(get_plan, req.user),
			create_account_plan,
			update_account_plan
		],
		function(err, account_plan){
			if(err) res.throw_error(err, 503);
			else res.send({"id": account_plan._id});
		}
	);
}
exports.get_plan = function(req,res) {
	AccountsPlans.findOne({account_id: req.params.account_id, active: true}, function (err, account_plan){
		if(err) res.throw_error(err, 503);
		else if(!account_plan) res.throw_error(null, 404);
		else {
			res.send(account_plan.returnJSON());
		}
	});
};

File: api/payparrot_api/controllers/accounts.py
# -*- coding: utf-8 -*-
import os, sys
from bson.objectid import ObjectId
from datetime import date, datetime
from hashlib import sha1

from dateutil.relativedelta import relativedelta
from bottle import route, request, response, static_file, SimpleTemplate, template, view, redirect

from payparrot_api.libs.exceptions import UnauthorizeException
from payparrot_dal import Accounts, AccountsSessions, AccountsPlans

#Move away from here

@route('/notifications', method="POST")
def callback():
    print request.forms.__dict__
    print request.query.__dict__
    return {}

@route('/login')
def callback():
    return static_file('login.html', os.path.join(os.path.abspath(os.path.dirname(__file__)), '../../../public/'))    


@route('/signup', method="POST")
def callback(db):
    params = {'startup':'','name':'','email':'','password':''}
    params.update(request.forms)
    account = Accounts.findOne(db, {'email': params.get("email","")})
    status = False
    if not account:
        new_account = Accounts(db, params)
        new_account._data['password'] = sha1(new_account.salt+params.get('password')).hexdigest()
        new_account.insert()
        status = True
    params['status'] = status
    return template('signedup',params)

@route('/accounts/setup', method="GET")
def callback(db, secure=True):
    return template('steps',{'account_id': request.account.id})    

@route('/accounts/subscriptions', method="GET")
def callback(db, secure=True):
    return template('subscriptions',{'account_id': request.account.id})    


#Accounts

@route('/accounts', method="POST")
def create_account(db):
    params = request.json
    account = Accounts(db,request.json)
    account._data['password'] = sha1(account.salt+params.get('password')).hexdigest()
    account.insert()
    response.status = 201
    return {'id': str(account.id)}

@route('/accounts/:account_id', method="GET")
def get_account(account_id,db, secure = True):
    if account_id == 'me':
        account_id = request.account.id
    account = Accounts.findOne(db, account_id)
    filter_parrots = {'account_id': ObjectId(account_id), 'active': True}
    account._data['stats']['parrots_total'] = db.subscriptions.find(filter_parrots).count()
    from_ = datetime.today()
    from_ = datetime(from_.year, from_.month, from_.day)
    to_ = from_ + relativedelta(days=+1)
    filter_parrots['created_at'] = {'$gte': from_, '$lt': to_}
    account._data['stats']['parrots_today'] = db.subscriptions.find(filter_parrots).count()
    account._data['stats']['payments_total'] = db.payments.find({'account_id': ObjectId(account_id), 'success': True}).count()
    filter_payments = {'account_id': ObjectId(account_id), 'success': True}
    filter_payments['created_at'] = {'$gte': from_, '$lt': to_}
    account._data['stats']['payments_today'] = db.payments.find(filter_payments).count()
                    # 'payments_total': 0,
                    # 'payments_today': 0
    if account:
        return account.JSON()
    else:
        response.status = 404
        return {}

@route('/accounts/:account_id', method="PUT")
def get_account(account_id, db, secure=True):
    account = Accounts.findOne(db, account_id)
    if account:
        account.update(request.json)
        return account.JSON()
    else:
        response.status = 404
        return {}

@route('/accounts/:account_id/credentials', method="GET")
def get_credentials(account_id, db, secure=True):
    account = Accounts.findOne(db, account_id)
    if account:
        return account.credentials
    else:
        response.status = 404
        return {}

@route('/logged', method="GET")
def logged(db,secure = True):
    account = request.account
    account_plan = AccountsPlans.findOne(db, {'account_id': ObjectId(account.id)})
    if not account:
        redirect('/login')
        return
    if not account.setup:
        redirect('/accounts/setup')
        return
    # if not account_plan:
    #     redirect('/accounts/subscriptions')
    #     return
    redirect('/app.html')

@route('/login', method="POST")
def login(db):
    account_by_email = Accounts.findOne(db, {
        'email': request.forms.get('email'),
    })
    
    if account_by_email:
        account = Accounts.findOne(db, {
            'email': request.forms.get('email'),
            'password': sha1(account_by_email.salt+request.forms.get('password')).hexdigest()
        })
        if account:
            account_session = AccountsSessions(db, {'account_id': account.id})
            account_session.insert()
            response.set_cookie('sid', account_session.session_id, path='/', expires = account_session.expires)
            redirect('/logged')
            return
    raise UnauthorizeException()

@route('/logout', method="GET")
def logout():
    response.set_cookie('sid', '', path='/')
    redirect('/login')

@route('/apply', method="POST")
def apply(db):
    applicant = {
        'email': request.json.get("email",""),
        'name': request.json.get("name","")
    }
    if applicant['email'] and applicant['name']:
        user = db.potential_users.find_one({'email':applicant['email']})
        if not user:
            db.potential_users.insert(applicant)
    else:
        response.status = 404
        return {}